atomic_reactor/utils/resource_manager.py

```diff
@@ -0,0 +1,132 @@
+                    if slot == '0\n':
+                        state = 'free'
+                        prid = ''
+                        free_slots += 1
+                    elif slot == "":
+                        logger.warning("host: %s slot: %s has invalid slot state", host, i)
+                        state = 'unknown'
+                        prid = 'unknown'
+                    else:
+                        state = 'running'
+                        logger.debug(slot)
+                        slot_data = json.loads(slot)
+                        prid = slot_data.get("prid", "unknown")
+                    logger.debug("%s slot %s is %s", meta['hostname'], i, state)
+                    meta['slots'].append({'sid': i, 'state': state, 'prid': prid})
+                    i += 1
+                meta['free_slots'] = free_slots
+                return meta
+        except Exception as err:
+            logger.debug(err)
+            logger.error("Availiability check failed fo host: %s", host)
+            return False
+
+    def unlock(self, user, host, slot):
+        logger.debug('Unlocking')
+        retries = 0
+        while retries < 3:
+            cmd = ('ssh %s@%s -C4 "echo \'0\' > %sslot_%s.json"', user, host, ROOT_PATH, slot)
+            p = subprocess.Popen(cmd, shell=True)
+            ret = p.wait()
+            if ret == 0:
+                logger.info('Unlocked')
+                return True
+            else:
+                logger.debug('Unlocking failed, retrying')
+                retries += 1
+        return False
+
+    def lock(self, user, host, slot, prid):
+        logger.debug("Aquiring lock")
+        payload = json.dumps({"prid": prid, "locked": str(datetime.now())})
+        retries = 0
+        while retries < 3:
+            filename = os.path.join(ROOT_PATH, f'slot_{slot}.json')
+            _cmd = ('if grep -qw 0 %s ; then echo %s > %s ;else exit 1; fi',
+                    filename,
+                    quote(payload),
+                    filename
+                    )
+            cmd = 'flock {} -c {}'.format(filename, quote(_cmd))
+            remote_cmd = 'ssh {}@{} -C4 {}'.format(user, host, quote(cmd))
+            logger.debug(cmd)
+            p = subprocess.Popen(remote_cmd, shell=True, universal_newlines=True)
+            ret = p.wait()
+            if ret == 0:
+                logger.info('Locked')
+                return True
+            else:
+                logger.debug('Locking failed, retrying')
+                retries += 1
+        logger.info('Locking failed')
+        return False
+
+    def find(self, arch, hosts):
+        try:
+            with open(hosts, "r") as f:
+                hosts_list = json.loads(f)
+            resources = []
+            for host in hosts_list['hosts']:
+                user = host.get('username', False)
+                hostname = host.get('hostname', False)
+                if user and hostname:
```

Comment:
IMO, we should just make user, host and arch mandatory and validate it in schema